src/main/scala/com/fulcrumgenomics/rnaseq/AnnotateByIntervalList.scala

```diff
@@ -0,0 +1,246 @@
+    |By default if a read is enclosed by multiple intervals it will be counted for both intervals. This behavior
+    |can be changed using a different `OverlapAssociationStrategy`
+  """)
+class AnnotateByIntervalList
+(
+  @arg(flag = 'i', doc = "Mapped BAM file.") val input: PathToBam,
+  @arg(flag = 'l', doc = "Path to intervals") val intervals: PathToIntervals,
+  @arg(flag = 'o', doc = "Path to the output metrics file") val output: FilePath,
+  @arg(flag = 'R', doc = "The regular expression to use to classify the name of the interval") val nameRegex: Option[String] = None,
+  @arg(flag = 'u', doc = "The name of the intervals not in the given set of intervals") val unmatched: String = "unmatched",
+  @arg(flag = 'm', doc = "The minimum mapping quality for a read to be included.") val minMappingQuality: Int = 20,
+  @arg(flag = 'q', doc = "The minimum base quality for a base to be included.") val minBaseQuality: Int = 0,
+  @arg(flag = 'a', doc = "The strategy to determine which interval overlapping templates should be associated with")
+  val overlapAssociationStrategy: OverlapAssociationStrategy = OverlapAssociationStrategy.Duplicate
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(this.input)
+  Io.assertReadable(this.intervals)
+  Io.assertCanWriteFile(this.output)
+
+  override def execute(): Unit = {
+    val progress = ProgressLogger(logger, verb = "Processed", noun = "loci", unit = 1000000)
+    val reader = SamSource(this.input)
+    val dict = reader.dict
+
+    val (baseCounter, uniqueBaseCounter, templateCounter, uniqueTemplateCounter, overlapDetector) = {
+      val _baseCounter            = new SimpleCounter[String]()
+      val _uniqueBaseCounter      = new SimpleCounter[String]()
+      val _templateCounter        = new SimpleCounter[String]()
+      val _uniqueTemplateCounter  = new SimpleCounter[String]()
+      val intervals               = buildIntervals(dict.asSam)
+      intervals.foreach { interval => _baseCounter.count(interval.getName, 0) }
+
+      val _overlapDetector = OverlapDetector.create[Interval](intervals.iterator.toJavaList)
+      (_baseCounter, _uniqueBaseCounter, _templateCounter, _uniqueTemplateCounter, _overlapDetector)
+    }
+
+    // Templates are assigned to an interval together.
+    val iterator = reader.iterator
+    val records = iterator.filterNot(record => record.duplicate
+      || record.secondary
+      || record.supplementary
+      || !record.pf
+      || !record.isFrPair
+      || record.mapq < minMappingQuality)
+
+    // First we filter to ensure that we only pick one read in a template and then iterate over those reads to create
+    // an interval for each template. For single ended reads the interval is just the read.
+    val templateIntervals = records.filter(record => {
+      if (record.mateMapped) {
+        val equalStartTieBreaker =
+          if (Math.min(record.start, record.end) == Math.min(record.mateStart, record.mateEnd.getOrElse(Int.MaxValue))) {
```

Comment:
Won't `Math.min(record.start, record.end)` always be `record.start`?   Also, what if this record is unmapped?

I think what you want is this:
1. record is unmapped and (not paired or mate is unmapped) => ignore
2. record is single-end (fragment) => `start=record.start` and `end=record.end`
3. record is mapped and mate is unmapped => same as (2)
4. record is unmapped and mate is mapped => use `record.mateStart` and `record.mateEnd.getOrElse(throw an exception saying you need a mate cigar)`
5. record is mapped and mate is mapped and `record.start == record.mateStart` and `record.firstOfPair` => start is `record.start` and end is `max(record.end, record.mateEnd.getOrElse(exception))`.
6. this should never happen since you filtered them out

Reply:
I think this handles 2, 3, 4, 5
```
        // Single end read mateStart will be set to 0
        Math.min(record.start, record.mateStart),
        // Set mate end to math min if it isn't defined (single end read)
        Math.max(record.end, record.mateEnd.getOrElse(Int.MinValue))
```
I don't throw an exception on `record.mateEnd.getOrElse` since that will mess up single ended reads.

Also for #4 there is already a require() in the accessor method to make sure mate cigar is set.